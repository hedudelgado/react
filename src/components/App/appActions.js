import { HeaderInterface, InvalidSessionErrorCodeMap, InvalidSessionHelper } from 'mega-code';
import {
  getCountries,
  getNewsletterRegions,
  getNewsletterPreferences,
  getProfile,
  updateProfile,
  deleteCustomer,
} from '../../utils/api';

// actions
export const USER_LOGGED_IN_SUCCESSFULLY = 'USER_LOGGED_IN_SUCCESSFULLY';
export const COUNTRIES_RECEIVED = 'COUNTRIES_RECEIVED';
export const PROFILE_RECEIVED = 'PROFILE_RECEIVED';
export const REGIONS_RECEIVED = 'REGIONS_RECEIVED';
export const NEWSLETTER_RECEIVED = 'NEWSLETTER_RECEIVED';
export const SET_EDIT_VIEW = 'SET_EDIT_VIEW';
export const SET_NOTIFICATION_MESSAGE = 'SET_NOTIFICATION_MESSAGE';
export const PROFILE_UPDATED = 'PROFILE_UPDATED';
export const NEWSLETTER_UPDATED = 'NEWSLETTER_UPDATED';
export const RESET_NEWSLETTER_INFO = 'RESET_NEWSLETTER_INFO';
export const RESET_USER_INFO = 'RESET_USER_INFO';
// views
export const DEFAULT_VIEW = undefined;
export const USER_DETAILS_VIEW = 'USER_DETAILS_VIEW';
export const PASSWORD_VIEW = 'PASSWORD_VIEW';
export const PAYMENT_CARD_VIEW = 'PAYMENT_CARD_VIEW';
export const ADD_GUEST_VIEW = 'ADD_GUEST_VIEW';
export const REGULAR_GUESTS_VIEW = 'REGULAR_GUESTS_VIEW';
export const NEWSLETTER_VIEW = 'NEWSLETTER_VIEW';
export const ROOM_REQUIREMENTS_VIEW = 'ROOM_REQUIREMENTS_VIEW';
export const EXTRAS_PREFERENCES_VIEW = 'EXTRAS_PREFERENCES_VIEW';
export const MEMORABLE_WORD_VIEW = 'MEMORABLE_WORD_VIEW';
// errors
export const PROFILE_ERROR = 'PROFILE_ERROR';
export const PASSWORD_ERROR = 'PASSWORD_ERROR';
export const NEWSLETTER_ERROR = 'NEWSLETTER_ERROR';
export const PAYMENT_CARD_ERROR = 'PAYMENT_CARD_ERROR';
export const USER_DETAILS_ERROR = 'USER_DETAILS_ERROR';
export const REGULAR_GUESTS_ERROR = 'REGULAR_GUESTS_ERROR';
export const MEMORABLE_WORD_ERROR = 'MEMORABLE_WORD_ERROR';
export const ROOM_REQUIREMENTS_ERROR = 'ROOM_REQUIREMENTS_ERROR';
export const EXTRAS_PREFERENCES_ERROR = 'EXTRAS_PREFERENCES_ERROR';

const checkForInvalidSession = (error, config, dispatch) => {
  if (InvalidSessionHelper
    .isSessionExpired(error, config)) {
    dispatch({
      type: RESET_USER_INFO,
    });
    dispatch({
      type: RESET_NEWSLETTER_INFO,
    });
    HeaderInterface.userLogOut(HeaderInterface.LOG_OUT_REASONS.SESSION_EXPIRED);
  }
};

export const getSettings = (email, sessionId, business) => (dispatch, getState) => {
  getProfile(email, sessionId, business).then((profile) => {
    dispatch({
      type: PROFILE_RECEIVED,
      profile,
    });
  }).catch((error) => {
    checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
    dispatch({
      type: PROFILE_ERROR,
      error,
    });
  });
  getCountries()
    .then((countries) => {
      dispatch({
        type: COUNTRIES_RECEIVED,
        countries,
      });
    });
  if (getState().application.showEmailPreferences) {
    getNewsletterRegions()
      .then((regions) => {
        dispatch({
          type: REGIONS_RECEIVED,
          regions,
        });
      });

    getNewsletterPreferences(email)
      .then((newsletter) => {
        dispatch({
          type: NEWSLETTER_RECEIVED,
          newsletter,
        });
      })
      .catch(e => dispatch({
        type: NEWSLETTER_ERROR,
        error: e,
      }));
  }
};

export const updateUserProfile = payload => dispatch => updateProfile(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.ACCOUNT.invalidSession, dispatch);
  return Promise.reject(error);
});

export const deleteUserProfile = payload => dispatch => deleteCustomer(payload).catch((error) => {
  checkForInvalidSession(error, InvalidSessionErrorCodeMap.DELETE_PROFILE.invalidSession, dispatch);
  return Promise.reject(error);
});

export default {
  getSettings,
};
