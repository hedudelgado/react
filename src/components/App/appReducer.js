import {
  SET_NOTIFICATION_MESSAGE, SET_EDIT_VIEW, DEFAULT_VIEW,
} from './appActions';

const defaultState = {
  editViewName: DEFAULT_VIEW,
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case SET_EDIT_VIEW:
      return {
        ...state,
        editViewName: action.editViewName,
      };
    case SET_NOTIFICATION_MESSAGE:
      return {
        ...state,
        notificationMessage: action.notificationMessage,
      };
    default:
      return state;
  }
};
