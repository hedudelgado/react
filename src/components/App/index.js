export { default } from './App';

export { default as AppActions } from './appActions';

export { default as AppContainer } from './AppContainer';

export { default as modalReducer } from './modalReducer';
