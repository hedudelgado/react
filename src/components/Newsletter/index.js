export { default } from './Newsletter';

export { default as NewsletterContainer } from './NewsletterContainer';
