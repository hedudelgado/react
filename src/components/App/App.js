import React, { Component } from 'react';
import {
  array,
  object,
  func,
  bool,
  node,
} from 'prop-types';
import {
  PASSWORD_VIEW,
  PAYMENT_CARD_VIEW,
  USER_DETAILS_VIEW,
  ROOM_REQUIREMENTS_VIEW,
  EXTRAS_PREFERENCES_VIEW,
} from './appActions';
import { ComponentTogglerContainer } from '../ComponentToggler';
import UserProfile from '../UserProfile';
import Password from '../Password';
import UserProfileFormContainer from '../UserProfileForm';
import PasswordFormContainer from '../PasswordForm';
import { PaymentCardContainer } from '../PaymentCard';
import { RegularGuestsContainer } from '../RegularGuests';
import { NewsletterContainer } from '../Newsletter';
import Modal from '../../layout/Modal';
import PaymentCardFormContainer from '../PaymentCardForm';
import { DeleteCustomerContainer } from '../DeleteCustomer';
import RoomRequirements from '../RoomRequirements';
import RoomRequirementsFormContainer from '../RoomRequirementsForm';
import ExtrasPreferences from '../ExtrasPreferences';
import ExtrasPreferencesFormContainer from '../ExtrasPreferencesForm';
import BaseStyle from '../../layout/BaseStyle';
import MemorableWord from '../MemorableWord';
import MoveAccount from '../MoveAccount';
import {
  STAYER, SELF, managePermissionFlag, USER_LOGGED_IN_EVENT,
} from '../../config.json';
import ManagePermission from '../ManagePermission';

class App extends Component {
  constructor(props) {
    super(props);
    this.handleLoggedIn = this.handleLoggedIn.bind(this);
  }

  componentDidMount() {
    window.addEventListener('message', this.handleLoggedIn);
  }

  componentWillUnmount() {
    window.removeEventListener('message', this.handleLoggedIn);
  }

  async handleLoggedIn(event) {
    if (event.data.action === USER_LOGGED_IN_EVENT) {
      const { getSettings } = this.props;
      const { contactDetail: { email }, sessionId, business } = JSON.parse(event.data.profile);
      getSettings(email, sessionId, !!business);
    }
  }

  render() {
    const {
      countries,
      dictionary,
      profile,
      regions,
      newsletter,
      displayModal,
      handleModalClose,
      modalContent,
      userDataReceived,
      application: {
        showEmailPreferences,
      },
    } = this.props;

    const {
      business,
      business: {
        accessLevel = '',
        tetheredGuid = '',
      } = {},
    } = profile;

    const appReady = userDataReceived
      && (!showEmailPreferences || !!regions.length)
      && !!countries.length
      && (!showEmailPreferences || !!Object.keys(newsletter).length)
      && !!Object.keys(dictionary).length
      && !!Object.keys(profile).length;

    const showExtrasPreferences = !(accessLevel === SELF || accessLevel === STAYER);

    return (
      <div data-test="AccountSettings">
        {appReady
        && (
          <BaseStyle className="wb-typography app-container">
            <ComponentTogglerContainer
              viewName={USER_DETAILS_VIEW}
              view={UserProfile}
              edit={UserProfileFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PASSWORD_VIEW}
              view={Password}
              edit={PasswordFormContainer}
            />
            <ComponentTogglerContainer
              viewName={PAYMENT_CARD_VIEW}
              view={PaymentCardContainer}
              edit={PaymentCardFormContainer}
            />
            {managePermissionFlag && (
              <ManagePermission dictionary={dictionary} />
            )}
            <RegularGuestsContainer />
            {showEmailPreferences && (
              <NewsletterContainer />
            )}
            <ComponentTogglerContainer
              viewName={ROOM_REQUIREMENTS_VIEW}
              view={RoomRequirements}
              edit={RoomRequirementsFormContainer}
            />
            {showExtrasPreferences && (
              <ComponentTogglerContainer
                viewName={EXTRAS_PREFERENCES_VIEW}
                view={ExtrasPreferences}
                edit={ExtrasPreferencesFormContainer}
              />
            )}
            {!!tetheredGuid && (
              <MemorableWord />
            )}
            {business && !tetheredGuid && (
              <MoveAccount />
            )}
            {!business && (
            <DeleteCustomerContainer />
            )}
            {displayModal && <Modal data-test="Modal" display content={modalContent} handleClose={handleModalClose} />}
          </BaseStyle>
        )
      }
      </div>
    );
  }
}

App.propTypes = {
  countries: array.isRequired,
  dictionary: object.isRequired,
  profile: object.isRequired,
  getSettings: func.isRequired,
  displayModal: bool.isRequired,
  handleModalClose: func.isRequired,
  modalContent: node,
  newsletter: object.isRequired,
  regions: array.isRequired,
  userDataReceived: bool.isRequired,
  application: object.isRequired,
};

export default App;
