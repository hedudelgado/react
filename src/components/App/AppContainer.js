import { connect } from 'react-redux';
import { getSettings as getSettingsAction } from './appActions';
import { closeModal } from './modalActions';

import App from './App';

const mapStateToProps = ({
  modal: {
    display: displayModal,
    content: modalContent,
  },
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userProfile: {
    userDataReceived,
    profile,
  },
}) => ({
  profile,
  displayModal,
  modalContent,
  dictionary,
  config,
  countries,
  regions,
  newsletter,
  application,
  userDataReceived,
});

const mapDispatchToProps = dispatch => ({
  getSettings:
    (email, sessionId, business) => dispatch(getSettingsAction(email, sessionId, business)),
  handleModalClose: () => dispatch(closeModal()),
});

export default connect(mapStateToProps, mapDispatchToProps)(App);
