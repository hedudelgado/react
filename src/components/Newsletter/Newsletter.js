import React, { Component } from 'react';
import {
  array,
  object,
  func,
} from 'prop-types';
import Typography from '../../layout/Typography';
import BaseStyle from '../../layout/BaseStyle';
import Button from '../../layout/Button';
import Notification from '../../layout/Notification';
import { updateNewsletter } from '../../utils/api';
import { NEWSLETTER_ERROR, NEWSLETTER_VIEW } from '../App/appActions';
import { marketingRegions } from '../../config.json';

class Newsletter extends Component {
  constructor(props) {
    super(props);

    const { newsletter: { regionSubscriptions } } = props;

    const regionsSubscribed = regionSubscriptions.reduce((rSubscribed, region) => {
      if (region.subscribed) {
        rSubscribed.push(+region.regionId);
      }
      return rSubscribed;
    }, []);

    this.state = {
      regionSubscriptions,
      regionsSubscribed,
    };

    this.handleNewsletterSubmit = this.handleNewsletterSubmit.bind(this);
    this.handleCheckboxChange = this.handleCheckboxChange.bind(this);
  }

  handleCheckboxChange(e) {
    const { regionSubscriptions } = this.state;
    // The checkbox indicates the user NOT wanting to receive emails,
    // so when checked subscribed should be false
    const subscribed = !e.target.checked;
    const regionId = +e.target.attributes.region.value;
    const index = regionSubscriptions.findIndex(i => +i.regionId === regionId);

    regionSubscriptions[index] = {
      regionId,
      subscribed,
    };

    this.setState({ regionSubscriptions });
  }

  handleNewsletterSubmit(e) {
    e.preventDefault();

    const {
      contactDetails: {
        email: emailAddress,
        address: {
          countryCode: countryCodeFromProfile,
        },
        firstName: firstNameFromProfile,
        lastName: lastNameFromProfile,
      },
      newsletter: {
        countryCode = countryCodeFromProfile,
        firstName = firstNameFromProfile,
        lastName = lastNameFromProfile,
        sessionId,
      },
      updateNewsletterStore,
      setNotificationMessage,
    } = this.props;

    const { regionSubscriptions } = this.state;

    const payload = {
      emailAddress,
      regionSubscriptions,
      sessionId,
      firstName,
      lastName,
      countryCode,
    };

    updateNewsletter(payload)
      .then(() => {
        setNotificationMessage(NEWSLETTER_VIEW);
        updateNewsletterStore(payload);
      }).catch(() => {
        setNotificationMessage(NEWSLETTER_ERROR);
      });
  }

  render() {
    const {
      app: {
        notificationMessage,
      },
      dictionary,
      regions,
    } = this.props;

    const { regionsSubscribed } = this.state;

    const newsletterOptions = regions
      .filter((region) => {
        const currentId = +region.id;
        return marketingRegions.includes(currentId);
      })
      .map((region) => {
        const currentId = +region.id;
        const regionIdSelector = `newsletter-region-${currentId}`;

        return (
          <BaseStyle
            key={region.id}
            component="div"
            data-test={regionIdSelector}
            className="wb-checkbox wb-checkbox__white mb2"
          >
            <input
              id={regionIdSelector}
              name={regionIdSelector}
              region={currentId}
              type="checkbox"
              // The checkbox indicates the user NOT wanting to receive emails,
              // so it should display the opposite of subscribed status
              defaultChecked={!regionsSubscribed.includes(currentId)}
              onChange={this.handleCheckboxChange}
            />
            <label htmlFor={regionIdSelector} className="font-size--base">
              {dictionary[`newsletter.region.${currentId}`]}
            </label>
          </BaseStyle>
        );
      });

    return (
      <BaseStyle data-test="newsletter-component" component="section" className="component-container">
        {notificationMessage === NEWSLETTER_ERROR && (
          <Notification data-test="error-notification" priority={2} variant="error" ariaLive="polite" show>
            {dictionary['generic.error']}
          </Notification>
        )}
        {notificationMessage === NEWSLETTER_VIEW && (
          <Notification data-test="success-message" priority={2} variant="success" ariaLive="polite" show>
            {dictionary['update.success.message']}
          </Notification>
        )}
        <Typography component="h4" className="pi-color-greyscale-darkest">
          {dictionary['newsletter.title']}
        </Typography>
        <Typography mb="2" dangerouslySetInnerHTML={{ __html: dictionary['newsletter.body'] }} />
        <form name="newsletter-form" onSubmit={this.handleNewsletterSubmit}>
          <BaseStyle component="section" mb="3">
            {newsletterOptions}
          </BaseStyle>
          <Button data-test="update-newsletter" color="secondary" mt="2" type="submit" fullWidth>
            {dictionary['newsletter.button.save']}
          </Button>
        </form>
      </BaseStyle>
    );
  }
}

Newsletter.propTypes = {
  contactDetails: object.isRequired,
  app: object.isRequired,
  dictionary: object.isRequired,
  regions: array.isRequired,
  newsletter: object.isRequired,
  updateNewsletterStore: func.isRequired,
  setNotificationMessage: func.isRequired,
};

export default Newsletter;
