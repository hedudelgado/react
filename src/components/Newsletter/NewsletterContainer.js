import { connect } from 'react-redux';
import Newsletter from './Newsletter';
import { SET_NOTIFICATION_MESSAGE, NEWSLETTER_UPDATED } from '../App/appActions';

const mapStateToProps = ({
  app,
  dictionary,
  userProfile: {
    profile: {
      contactDetail: contactDetails,
    },
  },
  regions,
  newsletter,
  application,
}) => ({
  app,
  dictionary,
  contactDetails,
  regions,
  newsletter,
  application,
});

const mapDispatchToProps = dispatch => ({
  setNotificationMessage: messageName => dispatch({
    type: SET_NOTIFICATION_MESSAGE,
    notificationMessage: messageName,
  }),
  updateNewsletterStore: newsletter => dispatch({
    type: NEWSLETTER_UPDATED,
    newsletter,
  }),
});

export default connect(mapStateToProps, mapDispatchToProps)(Newsletter);
