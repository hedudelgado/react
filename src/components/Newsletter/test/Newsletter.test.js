import React from 'react';
import { shallow } from 'enzyme';
import Newsletter from '../Newsletter';
import { updateNewsletter } from '../../../utils/api';
import { NEWSLETTER_ERROR, NEWSLETTER_VIEW } from '../../App/appActions';

jest.mock('../../../utils/api', () => ({
  __esModule: true,
  updateNewsletter: jest.fn(() => Promise.resolve({})),
}));

describe('Newsletter Component', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      contactDetails: {
        email: 'emailFromProfile',
        address: {
          countryCode: 'countryCodeFromProfile',
        },
        firstName: 'firstNameFromProfile',
        lastName: 'lastNameFromProfile',
      },
      app: {},
      dictionary: {
        'newsletter.title': 'Email preferences',
        'newsletter.body': 'We would like to send you news and offers. If you would prefer not to receive these, please ensure this box is ticked.',
        'newsletter.region.1': 'I don\'t want to receive email market about Premier Inn hotels',
        'newsletter.button.save': 'Save changes',
      },
      regions: [
        {
          id: '1',
          description: 'UK & Ireland',
          active: true,
        },
        {
          id: '2',
          description: 'Dubai',
          active: true,
        },
        {
          id: '3',
          description: 'India',
          active: true,
        },
        {
          id: '4',
          description: 'hub by Premier Inn',
          active: true,
        },
        {
          id: '5',
          description: 'Germany',
          active: true,
        },
      ],
      newsletter: {
        sessionId: 'sessionIdFromNewsletter',
        firstName: 'firstNameFromNewsletter',
        lastName: 'lastNameFromNewsletter',
        countryCode: 'countryCodeFromNewsletter',
        emailAddress: 'fresh@mailinator.com',
        regionSubscriptions: [
          {
            regionId: 1,
            subscribed: true,
          },
          {
            regionId: 2,
            subscribed: false,
          },
          {
            regionId: 3,
            subscribed: true,
          },
        ],
      },
      updateNewsletterStore: jest.fn(),
      setNotificationMessage: jest.fn(),
    };

    wrapper = shallow(<Newsletter {...props} />);
  });

  describe('Renders Newsletter Component Elements', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the Newsletter options', () => {
      expect(wrapper.find('[data-test^="newsletter-region-"]')).toHaveLength(4);
    });

    it('should render the update newsletter options button', () => {
      expect(wrapper.find('[data-test="update-newsletter"]')).toHaveLength(1);
    });
  });

  describe('Newsletter Preferences Business logic', () => {
    it('should show checkboxes as un-checked for subscribed regions', () => {
      expect(wrapper.find('[data-test^="newsletter-region-1"]').dive().find('[name="newsletter-region-1"]')).not.toBeChecked();
    });

    it('should show checkboxes as checked for un-subscribed regions', () => {
      expect(wrapper.find('[data-test^="newsletter-region-2"]').dive().find('[name="newsletter-region-2"]')).toBeChecked();
    });
  });

  describe('Marketing newsletters form submission', () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('should execute Newsletter update success procedure', async () => {
      const updatedRegionValue = 1;
      const option = wrapper
        .find(`[data-test^="newsletter-region-${updatedRegionValue}"]`)
        .dive()
        .find(`[name="newsletter-region-${updatedRegionValue}"]`);

      const event = {
        target: {
          checked: true,
          attributes: {
            region: {
              value: updatedRegionValue,
            },
          },
        },
      };

      option.simulate('change', event);
      wrapper.find('[name="newsletter-form"]').simulate('submit', { preventDefault: () => {} });
      await updateNewsletter;

      expect(updateNewsletter).toHaveBeenCalledWith({
        ...props.newsletter,
        emailAddress: 'emailFromProfile',
        regionSubscriptions: [
          {
            regionId: updatedRegionValue,
            subscribed: !event.target.checked,
          },
          {
            regionId: 2,
            subscribed: false,
          },
          {
            regionId: 3,
            subscribed: true,
          },
        ],
      });
      expect(props.setNotificationMessage).toHaveBeenCalledTimes(1);
      expect(props.updateNewsletterStore).toHaveBeenCalledTimes(1);
    });

    it('should include personal details from newsletter in payload if available', async () => {
      const option = wrapper
        .find('[data-test^="newsletter-region-1"]')
        .dive()
        .find('[name="newsletter-region-1"]');

      const event = {
        target: {
          checked: true,
          attributes: {
            region: {
              value: 1,
            },
          },
        },
      };

      option.simulate('change', event);
      wrapper.find('[name="newsletter-form"]').simulate('submit', { preventDefault: () => {} });

      expect(updateNewsletter).toHaveBeenCalledWith({
        ...props.newsletter,
        emailAddress: 'emailFromProfile',
        sessionId: 'sessionIdFromNewsletter',
        countryCode: 'countryCodeFromNewsletter',
        firstName: 'firstNameFromNewsletter',
        lastName: 'lastNameFromNewsletter',
      });
    });

    it('should include personal details from profile in payload if newsletter details are not available', async () => {
      props.newsletter.firstName = undefined;
      props.newsletter.lastName = undefined;
      props.newsletter.countryCode = undefined;

      wrapper = shallow(<Newsletter {...props} />);

      const option = wrapper
        .find('[data-test^="newsletter-region-1"]')
        .dive()
        .find('[name="newsletter-region-1"]');

      const event = {
        target: {
          checked: true,
          attributes: {
            region: {
              value: 1,
            },
          },
        },
      };

      option.simulate('change', event);
      wrapper.find('[name="newsletter-form"]').simulate('submit', { preventDefault: () => {} });

      expect(updateNewsletter).toHaveBeenCalledWith({
        ...props.newsletter,
        emailAddress: 'emailFromProfile',
        sessionId: 'sessionIdFromNewsletter',
        firstName: 'firstNameFromProfile',
        lastName: 'lastNameFromProfile',
        countryCode: 'countryCodeFromProfile',
      });
    });

    it('should execute Newsletter update failure procedure', async () => {
      wrapper = shallow(<Newsletter {...props} />);
      const option = wrapper
        .find('[data-test^="newsletter-region-1"]')
        .dive()
        .find('[name="newsletter-region-1"]');

      const event = {
        target: {
          checked: true,
          attributes: {
            region: {
              value: 1,
            },
          },
        },
      };

      const error = new Error('err');
      updateNewsletter.mockImplementationOnce(() => Promise.reject(error));

      option.simulate('change', event);
      wrapper.find('[name="newsletter-form"]').simulate('submit', { preventDefault: () => {} });

      await Promise.all([updateNewsletter]).then(() => {
        expect(props.setNotificationMessage)
          .toHaveBeenCalledTimes(1)
          .toHaveBeenCalledWith(NEWSLETTER_ERROR);
        expect(props.updateNewsletterStore).not.toHaveBeenCalled();
      });
    });

    it('should display the success notification', () => {
      props.app.notificationMessage = NEWSLETTER_VIEW;
      wrapper = shallow(<Newsletter {...props} />);

      expect(wrapper.find('[data-test="newsletter-component"]').dive())
        .toContainExactlyOneMatchingElement('[data-test="success-message"]');
    });

    it('should display the error notification', () => {
      props.app.notificationMessage = NEWSLETTER_ERROR;
      wrapper = shallow(<Newsletter {...props} />);

      expect(wrapper.find('[data-test="newsletter-component"]').dive())
        .toContainExactlyOneMatchingElement('[data-test="error-notification"]');
    });
  });
});
