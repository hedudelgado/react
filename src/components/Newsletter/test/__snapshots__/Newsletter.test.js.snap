// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletter Component Renders Newsletter Component Elements should match snapshot 1`] = `
<BaseStyle
  className="component-container"
  component="section"
  data-test="newsletter-component"
>
  <Typography
    className="pi-color-greyscale-darkest"
    component="h4"
  >
    Email preferences
  </Typography>
  <Typography
    component="p"
    dangerouslySetInnerHTML={
      Object {
        "__html": "We would like to send you news and offers. If you would prefer not to receive these, please ensure this box is ticked.",
      }
    }
    mb="2"
  />
  <form
    name="newsletter-form"
    onSubmit={[Function]}
  >
    <BaseStyle
      component="section"
      mb="3"
    >
      <BaseStyle
        className="wb-checkbox wb-checkbox__white mb2"
        component="div"
        data-test="newsletter-region-1"
        key="1"
      >
        <input
          defaultChecked={false}
          id="newsletter-region-1"
          name="newsletter-region-1"
          onChange={[Function]}
          region={1}
          type="checkbox"
        />
        <label
          className="font-size--base"
          htmlFor="newsletter-region-1"
        >
          I don't want to receive email market about Premier Inn hotels
        </label>
      </BaseStyle>
      <BaseStyle
        className="wb-checkbox wb-checkbox__white mb2"
        component="div"
        data-test="newsletter-region-2"
        key="2"
      >
        <input
          defaultChecked={true}
          id="newsletter-region-2"
          name="newsletter-region-2"
          onChange={[Function]}
          region={2}
          type="checkbox"
        />
        <label
          className="font-size--base"
          htmlFor="newsletter-region-2"
        />
      </BaseStyle>
      <BaseStyle
        className="wb-checkbox wb-checkbox__white mb2"
        component="div"
        data-test="newsletter-region-3"
        key="3"
      >
        <input
          defaultChecked={false}
          id="newsletter-region-3"
          name="newsletter-region-3"
          onChange={[Function]}
          region={3}
          type="checkbox"
        />
        <label
          className="font-size--base"
          htmlFor="newsletter-region-3"
        />
      </BaseStyle>
      <BaseStyle
        className="wb-checkbox wb-checkbox__white mb2"
        component="div"
        data-test="newsletter-region-4"
        key="4"
      >
        <input
          defaultChecked={true}
          id="newsletter-region-4"
          name="newsletter-region-4"
          onChange={[Function]}
          region={4}
          type="checkbox"
        />
        <label
          className="font-size--base"
          htmlFor="newsletter-region-4"
        />
      </BaseStyle>
    </BaseStyle>
    <Button
      color="secondary"
      data-test="update-newsletter"
      fullWidth={true}
      href={null}
      mt="2"
      onClick={[Function]}
      submit={false}
      type="submit"
      variant="button"
    >
      Save changes
    </Button>
  </form>
</BaseStyle>
`;
